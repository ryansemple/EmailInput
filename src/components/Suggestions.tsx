import React, { PureComponent } from "react";

import Suggestion from "../components/Suggestion";

interface ISuggestionsProps {
	showSuggestions: boolean,
	emailSuggestions: string[],
	emailSuggestionClickEvent: any
}

const Suggestions = (props: ISuggestionsProps) => {
	return (
		<div
			className={`${props.showSuggestions ? "show" : ""} Email_Suggestions float_left`}>
			<ul>
				{props.emailSuggestions
				.map
				(
					(emailSuggestion: string, index: number) => 
					<Suggestion
						key={emailSuggestion}
						emailSuggestion={emailSuggestion}
						emailSuggestionClickEvent={props.emailSuggestionClickEvent}
					/>
				)}
			</ul>
		</div>   
	)
}

export default Suggestions;