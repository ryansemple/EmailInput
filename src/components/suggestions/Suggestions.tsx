import React, { useState, useEffect } from "react";
import Suggestion from "./Suggestion";
import { isAlphaNumericRegex } from "../../utility/Regex";
import { atCharacter } from "../../utility/String";
import { popularEmailDomains } from "../../utility/Email";

interface SuggestionsProps {
	setEmail: (email: string) => void,
	email: string
}

const EmailSuggestions = (props: SuggestionsProps) => 
{
	const { email, setEmail } = props;
	const [emailSuggestions, setEmailSuggestions] = useState([""]);
	const [showEmailSuggestions, setShowEmailSuggestions] = useState(false);

	const emailSuggestionsClickEvent = (suggestedEmail: string): void => 
	{
		setEmail(suggestedEmail);
	}

	const checkForDomainToMatchEmail = 
	(
		emailCharactersAfterAtCharacter: string, 
		popularEmailDomain: string,
		emailCharactersBeforeAtCharacter: string
	): string => 
	{
		let suggestedEmail: string = "";
		const popularEmailDomainCharacters: string[] =
			popularEmailDomain.split("");
		let popularEmailMatch: boolean = true;
		
		for (let i: number = 0; i < emailCharactersAfterAtCharacter.length; i++)
		{
			const currentEmailCharacterAfterAtCharacter: string = emailCharactersAfterAtCharacter[i];
			const currentPopularEmailDomainCharacter: string = popularEmailDomainCharacters[i];
			
			if 
			(
				currentPopularEmailDomainCharacter &&
				currentPopularEmailDomainCharacter !== currentEmailCharacterAfterAtCharacter
			)
			{
				popularEmailMatch = false;
				break;
			}
		}

		if (popularEmailMatch)
		{
			suggestedEmail = `${emailCharactersBeforeAtCharacter}@${popularEmailDomain}.com`;
		}

		return suggestedEmail;
	}

	useEffect(() => {

		const isAlphaNumeric = (input: string): boolean => 
		isAlphaNumericRegex.test(input);

		const setSuggestionsAfterEmailChanges = (): void =>
		{
			let newEmailSuggestions: string[] = [];
	
			for (let i: number = 0; i < popularEmailDomains.length; i++) 
			{
				let suggestedEmail: string = "";
				const popularEmailDomain: string = popularEmailDomains[i];
	
				if (!email.includes(atCharacter))
				{
						suggestedEmail = `${email}@${popularEmailDomain}.com`;
				}
	
				if(email.includes(atCharacter))
				{
					const emailSplitByAtCharacter: string[] = email.split(atCharacter);
					const emailCharactersAfterAtCharacter: string = emailSplitByAtCharacter[1];
					const emailCharactersBeforeAtCharacter: string = emailSplitByAtCharacter[0];
	
					if 
					(
						emailCharactersAfterAtCharacter &&
						!isAlphaNumeric(emailCharactersAfterAtCharacter)
					)
					{
							continue;
					}
					
					if (emailCharactersAfterAtCharacter.length === 0)
					{
						suggestedEmail = `${email}${popularEmailDomain}.com`;
					} 
					else 
					{
						suggestedEmail = checkForDomainToMatchEmail
						(
							emailCharactersAfterAtCharacter,
							popularEmailDomain,
							emailCharactersBeforeAtCharacter
						);
	
						if(suggestedEmail === "")
						{
							continue;
						}
					}
				}
					
				if (suggestedEmail && suggestedEmail !== "")
				{
					newEmailSuggestions.push(suggestedEmail);
				}     
			}
	
			setEmailSuggestions(newEmailSuggestions);
		}

		setSuggestionsAfterEmailChanges();

		if(email)
		{
			setShowEmailSuggestions(true);
		} 
		else 
		{
			setShowEmailSuggestions(false);
		}
	}, 
	[
		email,
		setShowEmailSuggestions
	]);
	
	return (
		<div className="Email_Suggestions full_width">
			{showEmailSuggestions &&
			<ul>
				{emailSuggestions.map((emailSuggestion: string) => 
					<Suggestion
						key={emailSuggestion}
						emailSuggestion={emailSuggestion}
						emailSuggestionClickEvent={emailSuggestionsClickEvent}
					/>
				)}
			</ul>
			}
		</div>
	)
}

export default EmailSuggestions;