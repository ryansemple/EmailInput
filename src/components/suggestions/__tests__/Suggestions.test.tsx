import React from "react";
import { render } from "@testing-library/react";
import Suggestions from "../Suggestions";

test("Renders Suggestions component correctly.", () => {
	const { container } = render
	(
		<Suggestions
			email="ryan@gmail.com"
			setEmail={() => {}}
		/>
	);
	
	expect(container).toMatchSnapshot();
});