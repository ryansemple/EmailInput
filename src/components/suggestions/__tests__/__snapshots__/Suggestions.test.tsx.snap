// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Suggestions component correctly. 1`] = `
<div>
  <div
    class="Email_Suggestions full_width"
  >
    <ul />
  </div>
</div>
`;
